libnopegl/src/glsl/helper_noise.glsl

```diff
@@ -0,0 +1,152 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+float u32tof32(highp uint x)
+{
+    uint y = (0x7FU << 23U) | (x >> 9U);
+    return uintBitsToFloat(y) - 1.0;
+}
+
+highp uint lowbias32(highp uint x)
+{
+    x ^= x >> 16;
+    x *= 0x7feb352dU;
+    x ^= x >> 15;
+    x *= 0x846ca68bU;
+    x ^= x >> 16;
+    return x;
+}
+
+uint hash(uvec2 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y));
+}
+
+uint hash(uvec3 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y ^ lowbias32(x.z)));
+}
+
+float random(uvec3 x)
+{
+    return u32tof32(hash(x));
+}
+
+/*
+ * References:
+ * - https://web.archive.org/web/20200618040237/https://mrl.nyu.edu/~perlin/paper445.pdf
+ * - https://web.archive.org/web/20180616011332/http://www.scratchapixel.com/lessons/procedural-generation-virtual-worlds%20/perlin-noise-part-2
+ */
+vec3 random_grad(uvec3 x)
+{
+    uint h = hash(x);
+    float r0 = u32tof32(0x00010001U * (h      & 0xffffU));
+    float r1 = u32tof32(0x00010001U * (h>>16U & 0xffffU));
+    float theta = acos(2.0 * r0 - 1.0);
+    float phi = ngli_tau * r1;
+    return vec3(cos(phi) * sin(theta), sin(phi) * sin(theta), cos(theta));
+}
+
+vec3 curve_quintic(vec3 t)
+{
+    return ((6. * t - 15.) * t + 10.) * t * t * t;
+}
+
+float noise_perlin(vec3 t, uint seed)
+{
+    vec3 i = floor(t);
+    vec3 f = fract(t);
+    uvec3 x = uvec3(i) + seed;
+
+    float y0 = dot(random_grad(x + uvec3(0, 0, 0)), f - vec3(0, 0, 0));
+    float y1 = dot(random_grad(x + uvec3(1, 0, 0)), f - vec3(1, 0, 0));
+    float y2 = dot(random_grad(x + uvec3(0, 1, 0)), f - vec3(0, 1, 0));
+    float y3 = dot(random_grad(x + uvec3(1, 1, 0)), f - vec3(1, 1, 0));
+    float y4 = dot(random_grad(x + uvec3(0, 0, 1)), f - vec3(0, 0, 1));
+    float y5 = dot(random_grad(x + uvec3(1, 0, 1)), f - vec3(1, 0, 1));
+    float y6 = dot(random_grad(x + uvec3(0, 1, 1)), f - vec3(0, 1, 1));
+    float y7 = dot(random_grad(x + uvec3(1, 1, 1)), f - vec3(1, 1, 1));
+
+    vec3 a = curve_quintic(f);
+    return mix(
+               mix(mix(y0, y1, a.x),
+                   mix(y2, y3, a.x),
+                   a.y),
+               mix(mix(y4, y5, a.x),
+                   mix(y6, y7, a.x),
+                   a.y),
+               a.z);
+}
+
+/* Adapted from https://en.wikipedia.org/wiki/Bicubic_interpolation */
+float bicubic(float a, float b, float c, float d, float t)
+{
+    mat4 m = 0.5 * mat4(
+        0.,  2.,  0.,  0.,
+       -1.,  0.,  1.,  0.,
+        2., -5.,  4., -1.,
+       -1.,  3., -3.,  1.
+    );
+    return dot(m * vec4(1., t, t*t, t*t*t), vec4(a, b, c, d));
+}
+
+float noise_block(vec3 t, uint seed)
+{
+    vec3 i = floor(t);
+    vec3 f = fract(t);
+    uvec3 x = uvec3(i) + seed;
+
+    float freq_x = random(uvec3(x.xy, seed));
+    float freq_y = random(uvec3(x.xy, seed << 16U));
+    float evolution = t.z * freq_x + freq_y;
```

Comment:
I'll need some explanations for that part if you don't mind

Reply:
Done.